docs/zkapps/o1js/foreign-fields.mdx

```diff
@@ -108,6 +106,214 @@ let y = SmallField.from(5n); // convert from bigint or number
+As usual, you can find more information about each method in the [API reference](../o1js-reference/classes/ForeignField).
+
+## Three kinds of foreign field
+
+If the basic usage examples look straightforward, here is where it gets a bit complicated.
+
+For each `ForeignField` class created with `createForeignField()`, there are actually three different variants. We call them _unreduced_, _almost reduced_ and _canonical_.
+
+You find the variants as static properties on the class; they are themselves classes:
+
+```ts
+let x = new Field17.Unreduced(0);
+let y = new Field17.AlmostReduced(0);
+let z = new Field17.Canonical(0);
+```
+
+Unreduced field elements just have the `ForeignField` type. For the other two variants, there are narrower base types that are common to each variant:
+
+```ts
+import { AlmostReducedField, CanonicalField } from 'o1js';
+
+y satisfies AlmostReducedField;
+z satisfies CanonicalField;
+```
+
+In the following, we explain when to use the different variants, and how to convert between them. You don't need to learn all of it, though: The type system will guide you to use the right variant in each situation.
+
+### Unreduced fields
+
+Most arithmetic operations return unreduced fields:
+
+```ts
+import assert from 'assert';
+
+let z = x.add(x);
+assert(z instanceof Field17.Unreduced);
+```
+
+In short, **unreduced** means that a value can be larger than the modulus.
+
+For example, if `x` has the value 16, it is valid for `x.add(x)` to contain the value 32. The addition is correct modulo 17, but doesn't guarantee a result smaller than 17.
+
+:::note
+
+Unreduced doesn't usually mean that the underlying witness is larger than the modulus. It just means that we have not _proved_ it to be smaller.
+A malicious prover _could_ make it larger, by slightly modifying their local version of o1js and creating a proof with that.
+
+:::
+
+Unreduced fields can be added and subtracted, but not multiplied or divided:
+
+```ts
+z.add(1).sub(x); // works
+
+assert((z as any).mul === undefined); // z.mul() is not defined
+assert((z as any).inv === undefined);
+assert((z as any).div === undefined);
+```
+
+### Almost reduced fields
+
+To do multiplication, you need almost reduced fields. You can convert to them by using `.assertAlmostReduced()`:
+
+```ts
+let zAlmost = z.assertAlmostReduced();
+assert(zAlmost instanceof SmallField.AlmostReduced);
+```
+
+Now you can do multiplication and division:
+
+```ts
+let zz = zAlmost.mul(zAlmost); // zAlmost.mul() is defined
+
+// but .mul() returns an unreduced field again:
+assert(zz instanceof SmallField.Unreduced);
+
+// zAlmost.inv() is defined, and returns an almost reduced field:
+assert(zAlmost.inv() instanceof SmallField.AlmostReduced);
+```
+
+It can be convenient to require almost reduced fields as inputs to your smart contract. To do so, create a class which can also serve as a type, and use its `.provable` property when passing to the state decorator:
```

Comment:
```suggestion
It can be convenient to require almost reduced fields as inputs to your smart contract. To do that, create a class that can also serve as a type and use its `.provable` property when passing to the state decorator:
```